martha/martha_v3.js

```diff
@@ -478,87 +489,130 @@ async function retrieveFromServers(params) {
+            if (bondProvider && accessId) {
+                try {
+                    const bondAccessTokenUrl = `${config.bondBaseUrl}/api/link/v1/${bondProvider}/accesstoken`;
+                    console.log(`Requesting Bond access token for '${url}' from '${bondAccessTokenUrl}'`);
+                    const accessTokenResponse = await apiAdapter.getJsonFrom(bondAccessTokenUrl, auth);
+                    accessToken = accessTokenResponse.token;
+                } catch (error) {
+                    throw new RemoteServerError(error, 'Received error contacting Bond.');
+                }
+            }
+
+            // Retrieve the accessUrl using the returned accessToken, even if the token was empty.
+            if (bondProvider && accessId) {
+                try {
+                    const httpsAccessUrl = generateAccessUrl(drsType, accessId);
+                    const accessTokenAuth = `Bearer ${accessToken}`;
+                    console.log(`Requesting DRS access URL for '${url}' from '${httpsAccessUrl}'`);
+                    accessUrl = await apiAdapter.getJsonFrom(httpsAccessUrl, accessTokenAuth);
+                } catch (error) {
+                    throw new RemoteServerError(error, 'Received error contacting DRS provider.');
+                }
+            }
         } catch (error) {
-            throw new RemoteServerError(error, 'Received error contacting Bond.');
+            // Just logging the error for now. Eventually, we'll want to remove this outer try.
+            console.warn('Ignoring error from fetching signed URL:', error);
         }
-    }
+    };
 
-    // Retrieve the accessUrl using the returned accessToken, even if the token was empty.
-    let accessUrl;
-    if (bondProvider && accessId) {
-        try {
-            const httpsAccessUrl = generateAccessUrl(drsType, accessId);
-            const accessTokenAuth = `Bearer ${accessToken}`;
-            console.log(`Requesting DRS access URL for '${url}' from '${httpsAccessUrl}'`);
-            accessUrl = await apiAdapter.getJsonFrom(httpsAccessUrl, accessTokenAuth);
-        } catch (error) {
-            throw new RemoteServerError(error, 'Received error contacting DRS provider.');
-        }
+    let ranOutOfTime;
+    const waitUntilTimeIsAlmostUp = async () => {
+        await delay(pencilsDownTime * 1000);
+        ranOutOfTime = true;
+    };
+
+    // For now, since we can still return native object URLs, we don't want to fail if it takes too
+    // long to fetch a signed URL. Eventually, we won't have native object URLs to fall back on so
+    // we'll have to throw an error, at which point we can remove all of this `Promise.race` stuff.
+    await Promise.race([
+        fetch(),
+        waitUntilTimeIsAlmostUp(),
+    ]);
+
+    if (ranOutOfTime && !hypotheticalErrorMessage) {
+        console.log('Ran out of time fetching a signed URL, but returning everything else we fetched instead of throwing an error.');
```

Comment:
I went with "everything else" because the request might not have asked for any fields from DRS, like only requesting an SA key... but now that I type that, I realize that we wouldn't hit this case unless we were trying to get a signed URL which implies that we fetched the rest of the DRS metadata, so 👍 .